lib/portage/elog/mod_custom.py

```diff
@@ -18,8 +39,49 @@ def process(mysettings, key, logentries, fulltext):
         mylogcmd = mysettings["PORTAGE_ELOG_COMMAND"]
         mylogcmd = mylogcmd.replace("${LOGFILE}", elogfilename)
         mylogcmd = mylogcmd.replace("${PACKAGE}", key)
-        retval = portage.process.spawn_bash(mylogcmd)
-        if retval != 0:
+        loop = asyncio.get_event_loop()
+        proc = portage.process.spawn_bash(mylogcmd, returnproc=True)
+        procs = _get_procs()
+        procs.append((proc, asyncio.ensure_future(proc.wait(), loop=loop)))
+        for index, (proc, waiter) in reversed(list(enumerate(procs))):
+            if not waiter.done():
+                continue
+            del procs[index]
+            if waiter.result() != 0:
+                raise portage.exception.PortageException(
+                    f"!!! PORTAGE_ELOG_COMMAND failed with exitcode {waiter.result()}"
+                )
+
+
+async def _async_finalize():
+    """
+    Async finalize is preferred, since we can wait for process exit status.
+    """
+    procs = _get_procs()
+    while procs:
+        proc, waiter = procs.pop()
+        if (await waiter) != 0:
+            raise portage.exception.PortageException(
+                f"!!! PORTAGE_ELOG_COMMAND failed with exitcode {waiter.result()}"
+            )
+
+
+def finalize():
```

Comment:
Sorry for my ignorance, but is `finalize()` actually used anywhere?

Reply:
It's called in the atexit hook that's registered in the elog_process function here:
```python
                if hasattr(m, "finalize") and not m.finalize in _elog_atexit_handlers:
                    _elog_atexit_handlers.append(m.finalize)
                    atexit_register(m.finalize)
```
In practice the atexit hook is invoked a little early though, just before the event loop is closed, when we call `portage.process.run_exitfuncs()` at the end of the AsyncioEventLoop _close_main method:
```python
    async def _close_main(self):
        await mod_custom._async_finalize()

        # Even though this has an exit hook, invoke it here so that
        # we can properly wait for it and avoid messages like this:
        # [ERROR] Task was destroyed but it is pending!
        if socks5.proxy.is_running():
            await socks5.proxy.stop()

        portage.process.run_exitfuncs()
```
For both mod_custom and the socks5 proxy, the atexit hooks refer to processes that _close_main cleans up. We could refactor _close_main  to wait for both mod_custom and the socks5 proxy simultaneously using asyncio.gather.

Reply:
Thanks for explaining.

Reply:
This made me realize that `portage.process.run_exitfuncs()` call actually triggers recursive calls to the AsyncioEventLoop close method via an [atexit hook](https://github.com/gentoo/portage/blob/dcbcac809213537afaa6b4f9822146a2e984f773/lib/portage/util/futures/_asyncio/__init__.py#L302). It's fine though because AsyncioEventLoop uses a _closing attribute to detect recursive calls and drop them.